### ISSUE TITLE
`UseOrEmpty` rule is not active

### ISSUE THREAD

**Issue by andrey-bolduzev** [detekt/detekt#3540]
## Expected Behavior
`UseOrEmpty` should report instances like `y ?: emptyList()`.

## Observed Behavior
Violating code doesn't trigger the rule.

## Steps to Reproduce
detekt 1.16.0

The rule is enabled in the config:

```yaml
style:
  UseOrEmpty:
    active: true
```

The snippet is taken directly from the detekt tests for the rule but for some reason it passes perfectly in an actual project.

```kotlin
fun test(x: List<Int>?) {
    val a = x ?: emptyList()
}
```

## Context
After updating to 1.16.0 and enabling some of the new rules (at least some of which worked and produced new reports), I noticed `UseOrEmpty` specifically wasn't getting triggered in a large project riddled with `?: empty*`s.

I can try to provide a sample project if this is not easily reproducible in other environments.

## Your Environment

* Version of detekt used: 1.16.0

**Comment by BraisGabin** [comment]
This rule requiters type solving, do you have it enabled?

**Comment by andrey-bolduzev** [comment]
Ah, thank you, missed this in one of the release notes. Then this bug is not relevant.

On another note though, I'm using a custom task and, from the docs, I couldn't figure out what exactly I need to pass to `classpath` to make it work...

**Comment by cortinico** [comment]
> I couldn't figure out what exactly I need to pass to `classpath` to make it work...

There is a page about this exactly: https://detekt.github.io/detekt/type-resolution.html

**Comment by andrey-bolduzev** [comment]
> > I couldn't figure out what exactly I need to pass to `classpath` to make it work...
> 
> There is a page about this exactly: [detekt.github.io/detekt/type-resolution.html](https://detekt.github.io/detekt/type-resolution.html)

Yep, thanks, I found as much. But there aren't many examples and this is gradle we're talking about.

Through quite a few experiments, the best I have come up with in a multi-module android project is:

```groovy
...
	jvmTarget = "1.8"
	classpath.from(fileTree(projectDir) {
		include("*/build/**/*.class")
	})
...
```

This results in 22k `UnusedImport` reports and I don't really know why. My best guess is this happens because the class files from libraries are not included in the matcher above. But I'm not sure how to proceed right now.

**Comment by cortinico** [comment]
> But there aren't many examples and this is gradle we're talking about.

Is there a reason why you're not using the `detektMain` and other generated tasks that already configure the classpath for you?

**Comment by andrey-bolduzev** [comment]
> > But there aren't many examples and this is gradle we're talking about.
> 
> Is there a reason why you're not using the `detektMain` and other generated tasks that already configure the classpath for you?

The only way to enable this is to apply the detekt gradle plugin to each gradle module in the project, right?

In the past I found it resulted in _much_ slower lint runs as compared to a single detekt task that runs once over all sources in the project:

```groovy
task detektAll(type: io.gitlab.arturbosch.detekt.Detekt) {
	parallel = true
	failFast = false
	autoCorrect = !project.hasProperty("ci")
	buildUponDefaultConfig = true
	dependsOn(":custom-detekt-rules:build")
	source = files(projectDir)
	config.from = project.file(".static/config/default-detekt-config.yml")
	include("**/*.kt")
	exclude("*/build/")
	reports {
		html {
			enabled = true
			destination = project.file("build/reports/detekt.html")
		}
		xml.enabled = false
	}
}
```

**Comment by cortinico** [comment]
> The only way to enable this is to apply the detekt gradle plugin to each gradle module in the project, right?

Yes

> In the past I found it resulted in much slower lint runs as compared to a single detekt task that runs once over all sources in the project:

I see. Doing a singe run without the classpath is obviously faster as you don't have type resolution enabled. I'm unsure you can have a single task and compute a classpath that would work for all the modules (i.e. the "sum" of all the classpaths) to enable type resolution.

**Comment by BraisGabin** [comment]
> In the past I found it resulted in much slower lint runs as compared to a single detekt task that runs once over all sources in the project

We fixed that issue in the gradle plugin sobre versions ago. Now the time is nearly the same. So you are safe.

**Comment by andrey-bolduzev** [comment]
@cortinico @BraisGabin thank you for the help and the tips, I will play with moving back to the per-module tasks.
